test/test_util.py

```diff
@@ -167,6 +168,61 @@ def test_row_1d_filter_a(self) -> None:
         with self.assertRaises(NotImplementedError):
             row_1d_filter(a1.reshape(1,2,5))
 
+    #---------------------------------------------------------------------------
+
+    def test_array_deepcopy_a1(self) -> None:
+        a1 = np.arange(10)
+        memo = {}
+        a2 = array_deepcopy(a1, memo)
+
+        self.assertNotEqual(id(a1), id(a2))
+        self.assertNotEqual(mloc(a1), mloc(a2))
+        self.assertFalse(a2.flags.writeable)
+        self.assertEqual(a1.dtype, a2.dtype)
+
+    def test_array_deepcopy_a2(self) -> None:
+        a1 = np.arange(10)
+        memo = {}
+        a2 = array_deepcopy(a1, memo)
+
+        self.assertNotEqual(id(a1), id(a2))
+        self.assertNotEqual(mloc(a1), mloc(a2))
+        self.assertIn(id(a1), memo)
+        self.assertEqual(memo[id(a1)].tolist(), a2.tolist())
+        self.assertFalse(a2.flags.writeable)
+
+
+    def test_array_deepcopy_b(self) -> None:
+        a1 = np.arange(10)
+        memo = {id(a1): a1}
+        a2 = array_deepcopy(a1, memo)
+
+        self.assertTrue(mloc(a1) == mloc(a2))
+
+
+    def test_array_deepcopy_c1(self) -> None:
+        mutable = [np.nan]
+        memo = {}
+        a1 = np.array((None, 'foo', True, mutable))
+        a2 = array_deepcopy(a1, memo)
+
+        self.assertTrue(id(a1) != id(a2))
+        self.assertTrue(mloc(a1) != mloc(a2))
+        self.assertTrue(id(a1[3]) != id(a2[3]))
+        self.assertFalse(a2.flags.writeable)
```

Comment:
```suggestion
        self.assertNotEqual(id(a1), id(a2))
        self.assertNotEqual(mloc(a1), mloc(a2))
        self.assertNotEqual(id(a1[3]), id(a2[3]))
        self.assertFalse(a2.flags.writeable)
```

Reply:
Sorry, I think you suggested these before: I should have gotten them already!